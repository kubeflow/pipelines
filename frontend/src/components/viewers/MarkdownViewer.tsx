/*
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import Viewer, { ViewerConfig } from './Viewer';
import { cssRaw } from 'typestyle';
import Markdown from 'markdown-to-jsx';

cssRaw(`
.markdown-viewer h1,
.markdown-viewer h2,
.markdown-viewer h3,
.markdown-viewer h4,
.markdown-viewer h5,
.markdown-viewer h6 {
position: relative;
  margin-top: 1em;
  margin-bottom: 16px;
  font-size: initial;
  font-weight: 700;
  line-height: 1.4;
}

.markdown-viewer h1,
.markdown-viewer h2 {
  padding-bottom: .3em;
  border-bottom: 1px solid #eee;
}

.markdown-viewer code,
.markdown-viewer pre {
  background-color: #f8f8f8;
  border-radius: 5px;
}

.markdown-viewer pre {
  border: solid 1px #eee;
  margin: 7px 0;
  padding: 7px;
}
`);

export interface MarkdownViewerConfig extends ViewerConfig {
  markdownContent: string;
}

interface MarkdownViewerProps {
  configs: MarkdownViewerConfig[];
  maxDimension?: number;
}

class MarkdownViewer extends Viewer<MarkdownViewerProps, any> {
  private _config = this.props.configs[0];

  public getDisplayName(): string {
    return 'Markdown';
  }

  public render(): JSX.Element | null {
    if (!this._config) {
      return null;
    }
    return (
      <div className='markdown-viewer'>
        <Markdown>{this._config.markdownContent}</Markdown>
      </div>
    );
  }
}

export default MarkdownViewer;
