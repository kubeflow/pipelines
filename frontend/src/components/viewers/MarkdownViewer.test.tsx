/*
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { mount } from 'enzyme';
import MarkdownViewer, { MarkdownViewerConfig } from './MarkdownViewer';
import { PlotType } from './Viewer';

describe('MarkdownViewer', () => {
  it('does not break on empty data', () => {
    const tree = mount(<MarkdownViewer configs={[]} />).getDOMNode();
    expect(tree).toMatchSnapshot();
  });

  it('renders some basic markdown', () => {
    const markdown = '# Title\n[some link here](http://example.com)';
    const config: MarkdownViewerConfig = {
      markdownContent: markdown,
      type: PlotType.MARKDOWN,
    };

    const tree = mount(<MarkdownViewer configs={[config]} />).getDOMNode();
    expect(tree).toMatchSnapshot();
  });

  it('sanitizes the markdown to remove XSS', () => {
    const markdown = `
      lower[click me](javascript&#x3a;...)lower
      upper[click me](javascript&#X3a;...)upper
    `;
    const config: MarkdownViewerConfig = {
      markdownContent: markdown,
      type: PlotType.MARKDOWN,
    };

    const tree = mount(<MarkdownViewer configs={[config]} />).getDOMNode();
    expect(tree).toMatchSnapshot();
  });

  it('returns a user friendly display name', () => {
    expect(MarkdownViewer.prototype.getDisplayName()).toBe('Markdown');
  });
});
