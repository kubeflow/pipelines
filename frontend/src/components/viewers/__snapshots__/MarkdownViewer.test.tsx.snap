// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkdownViewer does not break on empty data 1`] = `null`;

exports[`MarkdownViewer renders some basic markdown 1`] = `
<div
  class="markdown-viewer"
>
  <div>
    <h1
      id="title"
    >
      Title
    </h1>
    <p>
      <a
        href="http://example.com"
      >
        some link here
      </a>
    </p>
  </div>
</div>
`;

exports[`MarkdownViewer sanitizes the markdown to remove XSS 1`] = `
<div
  class="markdown-viewer"
>
  <pre>
    <code>
        lower[click me](javascript&#x3a;...)lower
  upper[click me](javascript&#X3a;...)upper
    </code>
  </pre>
</div>
`;
